### ISSUE TITLE
Segmentation fault in SysBus::read_8

### ISSUE THREAD

**Issue by xTibor** [michelhe/rustboyadvance-ng#113]
Affected game:
* Power Rangers - Ninja Storm (USA)

Backtrace:
```
Thread 1 "rustboyadvance-" received signal SIGSEGV, Segmentation fault.
0x000055555562e029 in <rustboyadvance_core::sysbus::SysBus as rustboyadvance_core::bus::Bus>::read_8 ()
(gdb) bt
#0  0x000055555562e029 in <rustboyadvance_core::sysbus::SysBus as rustboyadvance_core::bus::Bus>::read_8 ()
#1  0x00005555555fe94d in rustboyadvance_core::arm7tdmi::thumb::exec::<impl rustboyadvance_core::arm7tdmi::cpu::Core>::do_exec_thumb_ldr_str ()
#2  0x00005555555fec05 in rustboyadvance_core::arm7tdmi::thumb::exec::<impl rustboyadvance_core::arm7tdmi::cpu::Core>::exec_thumb_ldr_str_imm_offset ()
#3  0x000055555560054d in rustboyadvance_core::arm7tdmi::cpu::Core::step ()
#4  0x00005555555fa336 in rustboyadvance_core::gba::GameBoyAdvance::step ()
#5  0x00005555555fa1b6 in rustboyadvance_core::gba::GameBoyAdvance::frame ()
#6  0x00005555555be914 in rustboyadvance_sdl2::main ()
#7  0x00005555555c44c3 in std::rt::lang_start::{{closure}} ()
#8  0x00005555557c3ac8 in std::rt::lang_start_internal::{{closure}} () at src/libstd/rt.rs:52
#9  std::panicking::try::do_call () at src/libstd/panicking.rs:331
#10 std::panicking::try () at src/libstd/panicking.rs:274
#11 std::panic::catch_unwind () at src/libstd/panic.rs:394
#12 std::rt::lang_start_internal () at src/libstd/rt.rs:51
#13 0x00005555555c0042 in main ()
```

**Comment by xTibor** [comment]
I did a valgrind run, doesn't reveal more than the gdb backtrace. Valgrind also shows that the `yaml` crate used by the emulator is unsound, but that's out of the scope of this issue.

**Comment by xTibor** [comment]
It's crashing on trying to read BIOS address `00950209`.

**Comment by michelhe** [comment]
Could be open-bus related ?

**Comment by xTibor** [comment]
I tested it, open-bus seems to solve this one too with that new `if aligned > 0x3ffe` check.

**Comment by michelhe** [comment]
So I guess that it will be solved after #109 is merged

**Comment by xTibor** [comment]
Looks like there are still segfaults in other places after open-bus (crashing 16-bit writes to `iodev`). I think I'll investigate them later in an another issue.

**Comment by michelhe** [comment]
For starters, we could replace all `get_unchecked` to get sensible panics instead of segfaults.
(I only put these to avoid size checks at runtime)

**Comment by xTibor** [comment]
Or change them conditionally with `#[cfg(debug_assertions)]` so release builds doesn't get a performance hit.

**Comment by xTibor** [comment]
Well I found out why I was getting segfaults on the open-bus branch. I was an idiot and forgot to call `gba.init();` in my test bench. Everything works flawlessly when things are initialized. The constructor of `GameBoyAdvance` could benefit from some more defensive approach to initialization to avoid mistakes like that (builder pattern or calling `.init()` in `::new()`, etc.).

**Comment by michelhe** [comment]
Oh yeah, thats what I meant that the branch isn't polished.
Since moving to pointer territory I get reminded why I hate modern compilers.

Calling Gba::init (which basically passes its pointer to SysBus) inside Gba::new isn't working (at least in windows), since the compiler *moves* the Gba allocation after returning from `Gba::new`, so the pointer is made stale.

Is there a way to pass weak references to owned structs in rust that doesn't involve completely redesigning the project?

**Comment by michelhe** [comment]
fixed in c8d5094
